### ISSUE TITLE
Logging into C# Dev Kit causes 401 with individual license

### ISSUE THREAD

**Issue by kroppt** [microsoft/vscode-dotnettools#1512]
Type: <b>Bug</b>

1. Install C# Dev Kit
2. Open a solution with C# Dev Kit running
3. Log into C# Dev Kit with an individual license
4. See error logged under "C# Dev Kit": IdeBenefitsSource: Failed to fetch entitlements. Error: 'Error: Request to https://api.subscriptions.visualstudio.microsoft.com/Me/Entitlements/IDEBenefits?api-version=2023-03-26&caller=vscode failed with status code: 401 and response ""'
5. The language server does not function

Extension version: 1.10.18
VS Code version: Code 1.94.0 (d78a74bcdfad14d5d3b1b782f87255d802b57511, 2024-10-02T13:08:12.626Z)
OS version: Windows_NT x64 10.0.22631
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 9 7950X 16-Core Processor             (32 x 4492)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|undefined|
|Memory (System)|63.74GB (52.09GB free)|
|Process Argv|--crash-reporter-id a10ff85b-4c8a-4979-96cf-e564dec7cc0f|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscod805:30301674
binariesv615:30325510
vsaa593:30376534
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonnoceb:30805159
asynctok:30898717
pythonmypyd1:30879173
h48ei257:31000450
pythontbext0:30879054
accentitlementsc:30995553
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
dsvsc021:30996838
9c06g630:31013171
a69g1124:31058053
dvdeprecation:31068756
dwnewjupytercf:31046870
impr_priority:31102340
nativerepl1:31139838
refactort:31108082
pythonrstrctxt:31112756
flighttreat:31134774
wkspc-onlycs-t:31132770
wkspc-ranged-t:31151552
cf971741:31144450
autoexpandse:31146404
iacca2:31150323
5fd0e150:31146321

```

</details>

**Comment by Li-su0** [comment]
Try these methods here; they can help alleviate your urgent needs.[1384](#1384),

**Comment by kroppt** [comment]
> Try these methods here; they can help alleviate your urgent needs. [1384](https://github.com/microsoft/vscode-dotnettools/issues/1384)

Thank you for the suggestion. Yes, I worked with only the C# extension, which does not need login. I disabled C# Dev Kit until this problem is fixed.

**Comment by AyrtonAlbuquerque** [comment]
I'm also having this problem but on Ubuntu only. In my windows 11 setup it works fine but in the Ubuntu this error occurs.

**Comment by arunchndr** [comment]
@AndreyTretyak would you please take a look? Related is https://github.com/microsoft/vscode-dotnettools/issues/1384

**Comment by kroppt** [comment]
I just tested this with the latest VSCode 1.95.0, C# 2.50.27, and C# Dev Kit 1.11.14.

The error shows up in the C# Dev Kit logs, but the behavior is a bit different. The LSP functionality appears to be working, but it's not clear whether the C# extension is falling back to Omnisharp or sticking with C# Dev Kit (somehow). It seems like it's working with the C# Dev Kit LSP.

C# logs:
```none
Using dotnet configured on PATH
Dotnet path: C:\Program Files\dotnet\dotnet.exe
Activating C# + C# Dev Kit...
waiting for named pipe information from server...
[stdout] {"pipeName":"\\\\.\\pipe\\bab8c11a"}
received named pipe information from server
attempting to connect client to server...
client has connected to server
[Info  - 8:06:16 PM] [Program] Language server initialized
[Info  - 8:06:18 PM] [WorkspaceProjectFactoryService] Project --snip--.csproj loaded by C# Dev Kit
[Info  - 8:06:18 PM] [WorkspaceProjectFactoryService] Project --snip--.csproj loaded by C# Dev Kit
```

C# Dev Kit logs:
```none
Starting Spawn .NET server...
Starting opening a solution...
Starting processing the solution file "--snip--.sln" in Dev Kit server...
Using preinstalled .NET runtime at "C:\Program Files\dotnet\dotnet.exe" to load Dev Kit processes.
Found installed dotnet location "C:\Program Files\dotnet\dotnet.exe" to load projects.
.NET server started and IPC established in 695ms
Completed Spawn .NET server (1042ms)
Completed processing the solution file "--snip--.sln" in Dev Kit server (1458ms)
Starting restoring NuGet packages...
Selected configuration: <Default>, active configuration: Debug|Any CPU
Completed opening a solution (1468ms)
Starting command: "dotnet.exe" restore --snip--.sln --interactive...
Completed command: "dotnet.exe" restore --snip--.sln --interactive (539ms)
Completed restoring NuGet packages (542ms)
IdeBenefitsSource: Failed to fetch entitlements. Error: 'Error: Request to https://api.subscriptions.visualstudio.microsoft.com/Me/Entitlements/IDEBenefits?api-version=2023-03-26&caller=vscode failed with status code: 401 and response ""'
Starting Query for all projects...
Completed Query for all projects (42ms)
```

If it helps, I noticed the Microsoft Authentication logs show the following:
```none
2024-10-29 20:06:17.201 [info] Getting sessions for all scopes...
2024-10-29 20:06:17.201 [info] Got 5 sessions for all scopes...
2024-10-29 20:06:18.524 [info]  for undefined
2024-10-29 20:06:18.524 [info] [--snip--/.default VSCODE_TENANT:organizations email offline_access openid profile] Got 1 sessions
```

Even though things are currently "working", I'd rather keep this open until we know it's fixed.

**Comment by AndreyTretyak** [comment]
We have an ongoing problem with `https://api.subscriptions.visualstudio.microsoft.com` service accessibility due to increased security requirements, but there is fallback `https://app.vssps.visualstudio.com` that we are calling for getting license that supposed to work and provider correct license.

So having this error in logs should not be blocking in any way.

Are you seeing incorrect licensing displayed by C# Dev Kit?
Are there any problems caused aside of this message being in the logs?
Even with us failing to obtain license via both sources you should be able to fully use C# Dev Kit without any blockers.

**Comment by kroppt** [comment]
@AndreyTretyak 
> Are you seeing incorrect licensing displayed by C# Dev Kit?

When I click on the C# button in the bottom right, it says, "This software is licensed to --snip-- under an individual license." That is the correct license.

> Are there any problems caused aside of this message being in the logs?

No, not currently.

I don't know why it was failing initially and now functions. I don't know whether there was a bug that got fixed or if it was just "fricked". You may close it if you think that the bug does not exist.

@AyrtonAlbuquerque Are you still getting this issue?

**Comment by AyrtonAlbuquerque** [comment]
> [@AndreyTretyak](https://github.com/AndreyTretyak)
> 
> > Are you seeing incorrect licensing displayed by C# Dev Kit?
> 
> When I click on the C# button in the bottom right, it says, "This software is licensed to --snip-- under an individual license." That is the correct license.
> 
> > Are there any problems caused aside of this message being in the logs?
> 
> No, not currently.
> 
> I don't know why it was failing initially and now functions. I don't know whether there was a bug that got fixed or if it was just "fricked". You may close it if you think that the bug does not exist.
> 
> [@AyrtonAlbuquerque](https://github.com/AyrtonAlbuquerque) Are you still getting this issue?

I'm not. everything seems fine now, although i have not changed anything.

**Comment by webreidi** [comment]
@AndreyTretyak any update on this issue?

**Comment by vborovikov** [comment]
Is there a way to disable these 'benefits' or 'entitlements' or whatever it is completely? Otherwise, the extension keeps asking to sign in which is very annoying.

**Comment by markheath** [comment]
Any update on this? I've had this problem for about a month now

**Comment by AndreyTretyak** [comment]
@webreidi @arunchndr We are waiting on a partner team to resolve issue of their web service not supporting VS Code client id or VS Code to use client id with Native MSA, but this is not a highest priority for both teams right now, since customer impact limited to error message in the logs. We have a fallback service to get entitlements.

@vborovikov how often are asked to sign in? It's expected to be prompted once in couple weeks if you are not signed in. Using any MS account should remove prompting completely. If you are prompted more often it might be related to an error writing into your credential store or network configuration issues with firewall, proxy or vpn.

@markheath what problem are you referring to? This issue should not affect any functionality, it only writes error in the logs. Does C# Dev Kit fail to display correct license for you?

**Comment by markheath** [comment]
Yes, think you're right - have been battling a series of C# in VS Code issues on different computers over the past weeks (including this one https://github.com/microsoft/vscode-dotnettools/issues/1907). I went through one round of trying to fix this particular issue a couple of weeks back as it was an annoyance as I was screen recording some training materials and didn't want the distracting notification. The other issue is more of a showstopper, but only affects my other computer.

**Comment by vborovikov** [comment]
> [@vborovikov](https://github.com/vborovikov) how often are asked to sign in? It's expected to be prompted once in couple weeks if you are not signed in. Using any MS account should remove prompting completely. If you are prompted more often it might be related to an error writing into your credential store or network configuration issues with firewall, proxy or vpn.

The sign-in is always successful but the profile notification never disappears.

**Comment by coryalbert** [comment]
I'm also getting this issue with my install - it keeps showing the notification to sign in, and the C# Dev Kit Output keeps showing: `IdeBenefitsSource: Failed to fetch entitlements. Error: 'Error: Request to https://api.subscriptions.visualstudio.microsoft.com/Me/Entitlements/IDEBenefits?api-version=2023-03-26&caller=vscode failed with status code: 401 and response ""'` when I do.

Clicking on the C# icon on the bottom right part of the screen shows that it is licensed as an individual account, but this is highly annoying and confusing.

**Comment by darkvertex** [comment]
Also getting the issue. In the C# Dev Kit output log:
```
2025-06-23 10:12:49.765 [info] IdeBenefitsSource: Failed to fetch entitlements. Error: 'Error: Request to https://api.subscriptions.visualstudio.microsoft.com/Me/Entitlements/IDEBenefits?api-version=2023-03-26&caller=vscode failed with status code: 401 and response ""'
```

Logging in does nothing to make this go away and the little "1" stays in my Accounts widget permanently which is very distracting:
<img width="38" alt="Image" src="https://github.com/user-attachments/assets/10131e35-750f-4fdd-b11d-2f9fef836aeb" />

**Comment by Pahasara** [comment]
I'm facing the same issue.

**Comment by darkvertex** [comment]
@AndreyTretyak Sorry for the late response but restarting VSCode after sign in did not change anything for me.

**Comment by claudiaregio** [comment]
@AndreyTretyak Can we get a follow up on this please?

**Comment by StephenHodgson** [comment]
I haven't been able to log in since C# dev kit started.

```
IdeBenefitsSource: Failed to fetch entitlements. Error: 'Error: Request to https://api.subscriptions.visualstudio.microsoft.com/Me/Entitlements/IDEBenefits?api-version=2023-03-26&caller=vscode failed with status code: 401 and response
```

Happens across all my machines in vs code on windows. I've tried logging in with Github and live.com accounts. Nothing.
So I just use Visual Studio for now since VS Code is pretty much useless in C# atm w/o language server.

I'm already signed into VS Code with both my Live and GitHub accounts, why can't it just use those existing credentials?

**Comment by timheuer** [comment]
> pretty much useless in C# atm w/o language server.

@StephenHodgson I think we should ensure your login works on this, but just FYI there is no functionality that is prevented from not logging in, so you shouldn't be blocked from using the C# language server.

> Live and GitHub accounts, why can't it just use those existing credentials?

Yeah, amen. Unfortunately across extensions there is no shared keychain right now in VSCode

**Comment by ALiwoto** [comment]
Just a reminder that this is still not fixed and it's preventing users from using the extension.

**Comment by ghita** [comment]
with this there is still no go to definition (or decompiling an external dependency - seems no code navigation at all) - the usage of the extension is useless once I get this error, I can even delete the extension . Are there plans to address this or at least it's impact ? (I get IdeBenefitsSource: Failed to fetch entitlements. Error: 'Error: Request to https://api.subscriptions.visualstudio.microsoft.com/Me/Entitlements/IDEBenefits?api-version=2023-03-26&caller=vscode failed with status code: 401 and response ""')

**Comment by claudiaregio** [comment]
@ghita We are working on trying to find the right solution here - there are a lot of moving parts we are trying to account for and are discussing this internally. 

Are you saying without signing into the extension you are not getting language service features? Functionality should not be prevented from not logging in so I think you should file another issue with all the logs -> Command palette: "Developer: Open Extensions log folder" -> zip this entire folder and paste in the new issue

**Comment by ghita** [comment]
@claudiaregio thanks for the response ! Yes, the language services are affected sometimes but you might be right is not because of this. I think I saw this happening if there is no *.sln for the projects or when it's opened in a multi folder solution  a sln that does not reference all the projects from there. I'll get back to you if I figure this out ;-).

**Comment by StephenHodgson** [comment]
Also seeing language services affected as well. 

Jetbrains Resharper plugin finally available in VS Code, installed it and worked just fine on its own. Guess I'll be relying on a 3rd party extension instead lol

**Comment by michaelmccord** [comment]
I just encountered this issue. It was literally working yesterday...sat down at my machine today and it's not working. Exact error as depicted in the output window. I'm on a mac. I even tried deleting `Code Safe Storage` from the keychain and that still has not worked. This essentially makes C# dev kit dead in the water for me, I'm unable to use my subscription.

**Comment by michaelmccord** [comment]
> I just encountered this issue. It was literally working yesterday...sat down at my machine today and it's not working. Exact error as depicted in the output window. I'm on a mac. I even tried deleting `Code Safe Storage` from the keychain and that still has not worked. This essentially makes C# dev kit dead in the water for me, I'm unable to use my subscription.

I was able to solve this by upgrading to the pre-release version of C# dev kit. I noticed when I did so I also had some problems with it trying to locate an older version of the dotnet cli in the home brew folder. I checked that out and it seems that the powershell homebrew version will override your installed version of the cli. So I uninstalled powershell via homebrew then manually installed it with the *.pkg file. That seemed to force it to resolve to the correct version of the CLI and all is well now. I will note that it the logs still contain the 401 from the IDE benefits service but it seems to eventually resolve this within the latest C# dev kit pre-release version.

**Comment by brwilkinson** [comment]
Can't figure this one out either. No luck in getting it working for a few weeks now.
same error message
>2026-06-20 22:52:20.287 [error] IdeBenefitsSource: Failed to fetch entitlements. Error: 'Error: Request to https://api.subscriptions.visualstudio.microsoft.com/Me/Entitlements/IDEBenefits?api-version=2023-03-26&caller=vscode failed with status code: 401 and response ""'

- Was fighting this for a while. I just ran dotnet SDK 10 update and now it appears to work. This was a new install about a month ago, so not sure what happened there.

- Recommendation if anyone else hits this run the "C# Workspace Requirements"... i never ran it before i had the issue, however at least after I upgraded i ran it and it confirm all requirements are met. So that might be a good place to start if others hit this.

okay, no luck it came back
> 2026-06-21 14:18:46.548 [error] IdeBenefitsSource: Failed to fetch entitlements. Error: 'Error: Request to https://api.subscriptions.visualstudio.microsoft.com/Me/Entitlements/IDEBenefits?api-version=2023-03-26&caller=vscode failed with status code: 401 and response ""'

```log
2026-06-21 14:34:24.482 [info] C# Dev Kit deactivate() is now executing...
2026-06-21 14:34:38.359 [info] SDK install view suppressed by user setting
2026-06-21 14:34:38.393 [info] Default solution: 
2026-06-21 14:34:52.872 [info] Starting: Opening a solution
2026-06-21 14:34:52.874 [info] Starting: processing the solution file "e:\MCPClientServerTest\MCPClientServerTest.slnx" in C# Dev Kit server
2026-06-21 14:34:52.874 [info] Checking if the opened file is a solution filter: e:\MCPClientServerTest\MCPClientServerTest.slnx
2026-06-21 14:34:52.923 [info] Starting: Spawn .NET server
2026-06-21 14:34:52.923 [info] Using dotnet resolution provided by ".NET Install Tool"
2026-06-21 14:34:53.740 [info] Using existing .NET runtime at "C:\Program Files\dotnet\dotnet.exe" to load C# Dev Kit processes
2026-06-21 14:34:54.424 [info] .NET SDK found: "C:\Program Files\dotnet\dotnet.exe"
2026-06-21 14:34:54.428 [info] Workspace folder: "e:\MCPClientServerTest" requires SDK version: 10.0.301
2026-06-21 14:34:54.433 [info] ASP.NET runtime found, using ASP.NET runtimeconfig for project system host.
2026-06-21 14:34:54.506 [info] vs-green server PID: 8240
2026-06-21 14:35:01.588 [info] .NET server started and IPC established in 7151ms
2026-06-21 14:35:01.641 [info] Completed: Spawn .NET server (8741ms)
2026-06-21 14:35:03.709 [error] IdeBenefitsSource: Failed to fetch entitlements. Error: 'Error: Request to https://api.subscriptions.visualstudio.microsoft.com/Me/Entitlements/IDEBenefits?api-version=2023-03-26&caller=vscode failed with status code: 401 and response ""'
```

**Comment by JipingWang** [comment]
> Can't figure this one out either. No luck in getting it working for a few weeks now. same error message

@brwilkinson , after a few pokes around, maybe enable sync settings, it was finally dismissed the login message on profile icon, while still see 401 error message in Output window.
~~~~~~~~
2026-07-05 08:20:14.314 [info] SDK install view suppressed by user setting
2026-07-05 08:20:14.318 [info] Default solution: 
2026-07-05 08:20:14.608 [error] IdeBenefitsSource: Failed to fetch entitlements. Error: 'Error: Request to https://api.subscriptions.visualstudio.microsoft.com/Me/Entitlements/IDEBenefits?api-version=2023-03-26&caller=vscode failed with status code: 401 and response ""'
2026-07-05 08:20:15.092 [info] Starting: Opening a solution
2026-07-05 08:20:15.092 [info] Starting: collecting .csproj files in the workspace in C# Dev Kit server
2026-07-05 08:20:15.092 [info] Checking if the opened file is a solution filter: null
2026-07-05 08:20:15.105 [info] Starting: Spawn .NET server
2026-07-05 08:20:15.105 [info] Using dotnet resolution provided by ".NET Install Tool"
2026-07-05 08:20:15.146 [info] Using existing .NET runtime at "C:\Program Files\dotnet\dotnet.exe" to load C# Dev Kit processes
2026-07-05 08:20:15.279 [info] .NET SDK found: "C:\Program Files\dotnet\dotnet.exe"
2026-07-05 08:20:15.280 [info] Workspace folder: "c:\Users\MyUserName\source\repos\MyOrgName\MyGitRepo" requires SDK version: 10.0.301
2026-07-05 08:20:15.281 [info] ASP.NET runtime found, using ASP.NET runtimeconfig for project system host.
2026-07-05 08:20:15.293 [info] vs-green server PID: 17044
2026-07-05 08:20:17.559 [info] .NET server started and IPC established in 2277ms
2026-07-05 08:20:17.574 [info] Completed: Spawn .NET server (2476ms)
2026-07-05 08:20:19.692 [info] Completed: collecting .csproj files in the workspace in C# Dev Kit server (4602ms)
2026-07-05 08:20:19.716 [info] Starting: Restoring NuGet packages
2026-07-05 08:20:19.716 [info] Selected configuration: <Default>, active configuration: Debug|Any CPU
2026-07-05 08:20:19.716 [info] Completed: Opening a solution (4626ms)
2026-07-05 08:20:19.753 [info] Starting: command: "C:\Program Files\dotnet\dotnet.exe" restore c:\Users\MyUserName\AppData\Roaming\Code\User\workspaceStorage\924d1fa4954c801131b167352cd92651\ms-dotnettools.csdevkit\MyGitRepo.sln --interactive --nodereuse:false
2026-07-05 08:20:21.192 [info] Querying process information for pshost using descriptor: [object Object] and got {"ProcessId":26192,"ProcessPath":"c:\\Users\\MyUserName\\.vscode\\extensions\\ms-dotnettools.csdevkit-3.20.199-win32-x64\\components\\vs-green-server\\platforms\\win32-x64\\node_modules\\@microsoft\\visualstudio-code-servicehost.win32-x64\\Microsoft.VisualStudio.Code.ServiceHost.exe"}
2026-07-05 08:20:21.193 [info] Discovered Project System Host (PID 26192) for pshost.
2026-07-05 08:20:21.193 [info] Project System Host PID: 26192
2026-07-05 08:20:22.453 [info] Completed: command: "C:\Program Files\dotnet\dotnet.exe" restore c:\Users\MyUserName\AppData\Roaming\Code\User\workspaceStorage\924d1fa4954c801131b167352cd92651\ms-dotnettools.csdevkit\MyGitRepo.sln --interactive --nodereuse:false (2726ms)
2026-07-05 08:20:22.454 [info] Completed: Restoring NuGet packages (2738ms)

**Comment by glen-84** [comment]
Still reproduces on a much newer build than anything reported here so far: **C# Dev Kit 3.20.199**, VS Code 1.134.0, Linux, August 2026. Same error string, same endpoint.

More usefully, I think the logs here rule out the authentication token as the cause, which is where most of the suggested fixes in this thread have been aimed.

### The token is valid at the moment of the 401

Dev Kit's bundled `extension.js` declares a distinct scope per entitlement probe:

| Source | Scope requested | Endpoint |
|---|---|---|
| `IdeBenefitsSource` | `6f1cc985-85e8-487e-b0dd-aa633302a731/.default` | `api.subscriptions.visualstudio.microsoft.com/…/IDEBenefits` |
| `LicensingWebSource` | `499b84ac-1321-427f-aa17-267ca6975798/.default` | `app.vssps.visualstudio.com/_apis/Licensing/ClientRights/VisualStudio` |
| `EntitlementsSource` | `https://graph.microsoft.com/.default` | `…/api/GetMyEntitlements` |

So `6f1cc985-…/.default` is exactly the scope `IdeBenefitsSource` needs. Correlating the **Microsoft Authentication** output channel against the **C# Dev Kit** channel, that scope had a live session immediately before each failure:

```
11:03:49.137 [createSession] [6f1cc985-…/.default] returned session
11:03:49.420 [getSessions]   [6f1cc985-…/.default] returned 1 session(s)
11:03:50.448 [error] IdeBenefitsSource: Failed to fetch entitlements ... status code: 401

11:04:04.774 [getSessions]   [6f1cc985-…/.default] returned 1 session(s)
11:04:05.985 [error] IdeBenefitsSource: Failed to fetch entitlements ... status code: 401
```

A freshly minted token for the probe's own declared scope, rejected one second later, twice. That points at the entitlements service refusing the account rather than at a missing or stale credential — which would explain why signing out and back in, re-granting access, and reinstalling all fail to change anything.

The account here is a personal Microsoft account with no Visual Studio subscription — `my.visualstudio.com/Subscriptions` reports "Sorry, we couldn't find any subscriptions for you." So the likeliest reading is that `IDEBenefits` returns 401 for an account that has no entitlement to report, rather than 403 or an empty 200.

If that server behaviour is intended, the remaining question is a client-side one: `IdeBenefitsSource` treats a settled answer as a retryable error and logs it at error level on every solution load. No user action can change the outcome, so that seems like the wrong severity for an expected response.

### A separate failure that makes the logs look like an auth problem

Worth separating out, because it sent me down the wrong path first: the *other* Dev Kit scope was genuinely failing to acquire silently, with an unrelated cause.

```
[getAllSessionsForPca] [499b84ac-…/.default] [<account>] failed to acquire token silently, skipping account
{"errorCode":"invalid_grant","errorMessage":"AADSTS700082: The refresh token has expired due to inactivity ...","subError":"bad_token"}
```

An interactive sign-in only consents the scope being requested at that moment, so `LicensingWebSource` keeps failing on a stale refresh token while `IdeBenefitsSource` has a valid one. Two distinct faults that read as one in the log.

Compounding it, VS Code's Microsoft auth provider never prunes an account whose refresh token is terminally dead — `getAllSessionsForPca` catches every error, logs "skipping account" at `info`, and continues, with no branch on `errorCode`. The dead entry is retried on every `getSessions` for every scope, indefinitely. That part is a VS Code issue rather than a Dev Kit one; microsoft/vscode#291146 is the closest existing report. Signing out of the account entirely, rather than signing in over it, cleared that half: one fresh consent then restored sessions for all three scopes above. The `IdeBenefitsSource` 401 was completely unaffected by it, which is the cleanest evidence I have that the two are independent.

Unlike the opening report, the language server is **not** broken here — the solution loads, projects are attached by `WorkspaceProjectFactoryService`, and Dev Kit is fully functional. On this version the 401 appears to be purely cosmetic, logged on every solution load with no user-visible consequence.

Happy to supply the full unredacted logs if that would help.

_Drafted by Claude (Anthropic AI assistant)._

**Comment by vborovikov** [comment]
A workaround I found is this: click on that small C# button and then the Allow button, then sign in into your account. The sign-in badge will disappear.

<img width="964" height="726" alt="Image" src="https://github.com/user-attachments/assets/f99323b2-dc3b-4eb8-90bc-c4605610edab" />
